### PR TITLE
UI: Add a fullscreen mode

### LINKED ISSUE #9334: Transform "Fix Toolbar to Top" into "Focus Mode"
Inspired by the "Focus" mode that @mtias introduced in #8920, I'd like to propose we try a different kind of Focus Mode.

## Problem

A key feedback point we hear is that Gutenberg’s interface can be a little overwhelming. This often comes from users who more commonly focus on "writing" versus "building" their posts. They find the contextual block controls and block hover states to be distracting: When they're focused on writing, they don't necessarily want to think about blocks — they just want to write.

Oftentimes, this subset of users also miss the common "formatting toolbar at the top of the page" paradigm that's present in Google Docs, Microsoft Word, and the Classic Editor.

I think we can introduce an alternate editing mode that addresses both these concerns for them.

## Suggested Solution

We already have a "Fix Toolbar to Top" option that moves the contextual block toolbar to the top of the page. For the user I described above, this is already a step towards the interface they're used to. It's also a good first step to decluttering the writing interface — relocating heavy UI to a less-disctracting area of the screen.

I suggest we take that option further, and adapt it into a more complete "Focus Mode":

Blueprints

This new editing mode would consist of a collection of UI updates aimed at decluttering the interface so that the user can focus on writing their content. 

▶️ **Video demo:** https://cloudup.com/cMr22auRtXC 

## Details

Focus Mode would be activated via the "More" menu. To accomodate this new mode, I propose renaming the "Fix Toolbar to Top" option to "Focus Mode" and including this as a new "Writing" option:

Writing Option in Menu
_(Since this would leave "Show Tips" all alone under "Settings", I suggest moving it into the "Tools" section at the bottom)_

When users have this new mode active, the editor would include the following UI updates:

### 1. The block toolbar would be pinned to the top of the screen.
(This is an existing feature.)

Pinned to Top Screenshot

### 2. The editor would be full screen.
This is one of the highest impact changes, and would be the default for this mode. Users could exit out of full screen mode — and retain all other features of Focus Mode — via a new toggle in the upper left of the toolbar.

full-screen-smaller

### 3. Block outlines would be removed for both hover and selected states. 
I initially thought this change would be confusing, but (as a power user myself) I find it quite usable. Since this is an optional mode, and this is a high-impact change in terms of eliminating distractions, I'm all for it. 

Block Borders GIF

### 4. The block label would appear on a delay, and be toned down visually.
This label is less essential in this mode, but including it will help with wayfinding. (The delay aspect of this change is already in progress in #9197)

Block Label

### 5. Block mover + block options would also appear on a delay. 
(For non-selected blocks). When a block is selected, they'll appear just as quickly as they usually do. 

_Non-selected Blocks_
Block Mover + Options (Unselected Block)

_Selected Blocks (This is the same as our current behavior)_
Block Mover + Options (Selected Block)

---

In case you missed it above, here's a short video demo to convey how these changes work in practice:

▶️  https://cloudup.com/cMr22auRtXC 

## FAQ

I foresee a few likely questions to this approach, so I'll try to address them in advance: 

- **How does this relate to the "Focus Mode" in #8920?** These can work together. The pattern of focusing in on a single block is compatible with both writing modes, so we could either allow it to be available in both modes, or we could make it an add-on feature for this Focus mode.
- **There are acessibility issues with not showing block borders.** That's very likely true. I'd reinforce the fact that this is a non-default, opt-in mode. If block borders and contextual controls are important to your use of Gutenberg, the default option will still be available. This direction would _not_ take that away. 
- **Nested blocks will still require borders in this mode**. Yes, maybe they will! We'll need to address this, and I think it's reasonable to include some sort of minimal border in this case. 
- **What about mobile?** We currently don't offer the "Fix Toolbar to Top" option on small screens, due to a [Safari issue noted here](https://github.com/WordPress/gutenberg/issues/7479#issuecomment-410988762). That will still be a problem for us. The other enhancements in this mode (full screen, hover state changes, etc.) would have little to no effect on mobile. For those reasons I suggest we limit this to larger screens only at this point. 

Looking forward to thoughts and reactions. 🙂

### PR DESCRIPTION
**Pull request by youknowriad**
closes #9334

This PR is the second part of #9334 and adds a fullscreen mode.
The CSS contains some "hacks" to allow switching modes without refreshing the page but I think it's fine.

### CODE REVIEW

File: edit-post/assets/stylesheets/_mixins.scss [hunk 1]
```diff
@@ -272,6 +272,10 @@
 			margin-left: -18px;
 		}
 	}
+
+	body.is-fullscreen-mode #{$selector} {
+		left: 0 !important;
```
Comment (youknowriad) [inline edit-post/assets/stylesheets/_mixins.scss]: I've used important to avoid overriding all the difference use-case above. It just seemed right :)

**Comment by kjellr** [comment]
This is looking awesome  — I really like how fullscreen feels. 

While using this, I clicked the `x` and expected it to just turn off full screen. 😄 Considering I should've known what to expect here, that seems like a good indication that we should make that button clearer. Maybe we should use the "close" button in the upper left, instead of the `x` icon:



I'm interested to hear what others think.

**Comment by youknowriad** [comment]
> While using this, I clicked the x and expected it to just turn off full screen. 😄 Considering I should've known what to expect here, that seems like a good indication that we should make that button clearer. 

I agree it's not perfect, but the current behavior is very similar to the customizer which is already a familiar fullscreen page for WordPressers

**Comment by youknowriad** [comment]
Damn buttons :P

**Comment by jasmussen** [comment]
Nice work, this is working better than I would've expected, actually. GIF:

fullscreen

The X for close is working well, and is used in other places as well. It's a bit curious, because it feels _very right_ to palace the X on the left here. But in other places, we place it on the right. Is this an inconsistency we want to look at? If yes, I think I'd prefer using a plain "Close" button rather than moving the X to the right, simply because because "More" menus should always be rightmost as they are to indicate "overflow" of buttons to the left of it, in an LTR reading direction. 

I suppose an exception could be made, but I'm not sure if it would look weird with the sidebar close button being stacked below. Though we could decide to hide the sidebar close button on the desktop breakpoint, since when it's a sidebar it's easily toggled using the cog.

The "back to where you were" isn't working quite as well as I thought it would. I think it might work better if it always took you to the same place — like the posts list. 

If the "Add new post" button in the toolbar was the primary interface (or only) for creating new posts, "back to where you were" would probably work better.

**Comment by youknowriad** [comment]
> The "back to where you were" isn't working quite as well as I thought it would. I think it might work better if it always took you to the same place — like the posts list.

I updated this as suggested.

File: edit-post/assets/stylesheets/main.scss [hunk 2]
```diff
@@ -105,11 +105,19 @@ body.gutenberg-editor-page {
 		bottom: 0;
 		left: 0;
 		min-height: calc(100vh - #{ $admin-bar-height-big });
+
+		body.is-fullscreen-mode & {
+			min-height: calc(100vh);
```
Comment (afercia) [inline edit-post/assets/stylesheets/main.scss]: is `calc()` necessary in this case?

File: edit-post/assets/stylesheets/main.scss [hunk 3]
```diff
@@ -105,11 +105,19 @@ body.gutenberg-editor-page {
...
+		}
 	}
 
 	// The WP header height changes at this breakpoint
 	@include break-medium {
 		min-height: calc(100vh - #{ $admin-bar-height });
+
+		body.is-fullscreen-mode & {
+			min-height: calc(100vh);
```
Comment (afercia) [inline edit-post/assets/stylesheets/main.scss]: is `calc()` necessary in this case?

File: edit-post/components/header/fullscreen-mode-close/index.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	if ( ! isActive ) {
+		return null;
+	}
+
+	return (
+		<Toolbar className="edit-post-fullscreen-mode-close__toolbar">
+			<IconButton
+				icon="no-alt"
+				href={ addQueryArgs( 'edit.php', { post_type: postType } ) }
+				label={ __( 'Close' ) }
```
Comment (afercia) [inline edit-post/components/header/fullscreen-mode-close/index.js]: Since this button is a link, and communicated as such to assistive technologies users, I'd rather consider to use a label that communicates the link destination instead of an action like "Close", which is more appropriate for buttons. RIght now, the link navigates back to the posts list, and it should communicate that 🙂

**Comment by afercia** [comment]
Nice job, also from an a11y perspective (but please see the comment about the "Close" link):
- the main H1 is preserved, and the headings hierarchy is correct
- the ARIA landmarks are preserved
- the `Ctrl+backtick` shortcut to navigate the main regions works
- the Tab sequence makes sense and there are no hidden tab stops

🎉

File: edit-post/components/header/fullscreen-mode-close/index.js (hunk 4 above)
Comment (afercia) [reply edit-post/components/header/fullscreen-mode-close/index.js]: It would work for me 👍 Yeah... I know about the Customizer link 🙈

**Comment by afercia** [comment]
And with the sidebar closed, it looks really nice! Visually, loving it 😍 

screen shot 2018-09-04 at 18 30 15

Have you considered to add a keyboard shortcut to toggle fullscreen mode? (yeah, new challenge to find a non-conflicting shortcut!)

**Comment by youknowriad** [comment]
I decided to use the `view_items` post type label for this button as it seemed the most appropriate label.

**Review by jasmussen — APPROVED** [review]
Technically this seems to be working very solidly.

There seems to be wide agreement among the reviewers here that this has value and should be tested. I happen to agree with that. 

It also adds value to laptops that have limited screen real estate, like 11 inch laptops, presumably tablets as well. As such, I think we should get this in and iterate.

**Comment by mtias** [comment]
Love this, thanks for the quick implementation!

**Comment by ktmn** [comment]
This is the best thing ever, thanks for making it!

**Comment by SNaushadS** [comment]
Thank you for all who worked on getting this done, I needed it badly.

I have some feedback on exiting/closing the fullscreen mode.

TLDR: X at the top left gave the affordance that i can get Out of the Full Screen mode. It should either close the mode or be removed.

What did i do:
1. Clicked on the hamburger menu.
2. Clicked on Full Screen mode.
   Full Screen Mode was now active, **Awesome**. It’s spacious and helps me focus on writing.
Then i wanted to get Out of the fullscreen mode. So i tried to find the way out.
3. Found the X at the top left and clicked on it.

**My expectation:** It should’ve got me out of fullscreen mode.
4. Instead i was provided warning by the browser if i want to Stay or Leave. Doesn’t match with user expectation. 
as already observed by @kjellr 
5. Then i tried to find out the “designed” way to close/exit Fullscreen Mode
fullscreengutenberg
Also, the mouse travel from left edge to right edge is a bit painful.
6. Fullscreen had a right tick next to it, clicked it and i was out of the fullscreen mode.

_Comments_
- Fullscreen behavior is typically that of a Modal.
- Hence the general expectation would be that the X either on the top-right or top-left would close/exit/take user out of the Mode.

_Suggestions:_
Let 
- The X could either allow user to exit the fullscreen mode (as per general interaction expectations)
- Or be removed and let user exit the mode only from hamburger menu by unticking.

Let me know how further i can help.

**Comment by aduth** [comment]
To @SNaushadS 's feedback, there's an iteration being proposed / discussed in #9838

**Comment by ktmn** [comment]
Looks like `wp.components.Modal` has a bug in fullscreen mode.



Above is the modal in normal mode. Below is the modal in fullscreen mode and the CSS rule causing it.

**Comment by jasmussen** [comment]
@ktmn good catch. I will take a look — I need to look at the fullscreen mode regardless as when an editor style applies a colored background, a bottom padding offsets this.

**Comment by jasmussen** [comment]
The issue you reported, @ktmn, will be fixed either by https://github.com/WordPress/gutenberg/pull/9973 or https://github.com/WordPress/gutenberg/pull/9900. They both address it.
